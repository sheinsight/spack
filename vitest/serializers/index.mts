/**
 * 统一的快照序列化器管理
 * 按优先级组合各种专用序列化器
 */

import type { SnapshotSerializer } from 'vitest';
import { pathNormalizerSerializer } from './path-normalizer.mts';
import { duplicateDependencySerializer } from './duplicate-dependency.mts';
import { errorsArraySerializer } from './errors-array.mts';
import { generalArraySerializer } from './general-array.mts';

/**
 * 序列化器优先级列表
 * 越前面的优先级越高
 */
const serializers: SnapshotSerializer[] = [
  pathNormalizerSerializer,  // 路径标准化应该最优先
  duplicateDependencySerializer,
  errorsArraySerializer,
  generalArraySerializer,
];

/**
 * 统一快照序列化器
 * 按优先级匹配并应用相应的序列化逻辑
 */
const unifiedSerializer: SnapshotSerializer = {
  test: (val: unknown): boolean => {
    // 检查是否有任何序列化器可以处理这个值
    return serializers.some((serializer) => serializer.test(val));
  },

  serialize: (val: any) => {
    // 按优先级应用序列化逻辑
    for (const serializer of serializers) {
      if (serializer.test(val)) {
        return (serializer as SnapshotSerializer).serialize(val);
      }
    }

    // 默认情况，不应该到达这里
    return JSON.stringify(val, null, 2);
  },
};

export default unifiedSerializer;
