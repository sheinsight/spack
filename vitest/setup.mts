process.env.NODE_ENV = 'test';
process.env.API_URL = 'http://localhost:3000';
process.env.RSPACK_BINDING = `${process.cwd()}/crates/binding`;
