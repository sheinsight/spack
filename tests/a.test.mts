import { test, expect } from 'vitest';
import { rspack, experiments } from '@rspack/core';
import path from 'node:path';
import * as binding from '@shined/spack-binding';

test('should sayHi correctly', async () => {
  const compiler = rspack({
    entry: {
      main: path.resolve(__dirname, 'fixtures/base/src/index.ts'),
    },
    output: {
      path: path.resolve(__dirname, 'fixtures/base/dist'),
      filename: 'bundle.js',
    },
  });
  const { promise, resolve } = Promise.withResolvers();

  compiler.run((err, stats) => {
    expect(err).toBeNull();
    expect(stats?.hasErrors()).toBe(false);
    expect(stats?.hasWarnings()).toBe(false);
    resolve(null);
  });

  await promise;
});

test('should sayHi correctly1', async () => {
  binding.registerCaseSensitivePathsPlugin();
  const CaseSensitivePathsPlugin = experiments.createNativePlugin(
    binding.CustomPluginNames.CaseSensitivePathsPlugin,
    (opt) => ({ ...opt })
  );
  const plugin = new CaseSensitivePathsPlugin({});
  const compiler = rspack({
    entry: {
      main: path.resolve(__dirname, 'fixtures/case_sensitive/src/index.ts'),
    },
    output: {
      path: path.resolve(__dirname, 'fixtures/case_sensitive/dist'),
      filename: 'bundle.js',
    },
    resolve: {
      extensions: ['.ts', '.js'],
    },
    plugins: [plugin],
  });
  const { promise, resolve } = Promise.withResolvers();

  compiler.run((err, stats) => {
    if (err) {
      resolve(true);
    }

    if (stats?.hasErrors()) {
      const json = stats?.toJson({
        errors: true,
      });
      resolve(json.errors);
    }

    resolve(false);
  });

  const result = await promise;
  expect(result).toMatchSnapshot();
});
