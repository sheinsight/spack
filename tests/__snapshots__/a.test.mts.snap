// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should sayHi correctly1 1`] = `
[
  {
    "code": "case mismatch",
    "details": undefined,
    "message": "  × Can't resolve "./Hello" in "<ROOT>/tests/fixtures/case_sensitive/src/index.ts". (case mismatch)
   ╭─[<ROOT>/tests/fixtures/case_sensitive/src/index.ts:1:22]
 1 │ import { hello } from './Hello';
   ·                       ────┬────
   ·                           ╰── Path case mismatch
 2 │ 
 3 │ hello();
   ╰────
  help: Fix the case of file paths to ensure consistency in cross-platform builds.
        It may work fine on macOS/Windows, but will fail on Linux.
",
    "moduleId": undefined,
    "moduleIdentifier": "<ROOT>/tests/fixtures/case_sensitive/src/index.ts",
    "moduleName": "./tests/fixtures/case_sensitive/src/index.ts",
    "moduleTrace": [],
    "stack": undefined,
  },
]
`;
